runtime/gc_base/GCExtensions.cpp

```diff
@@ -241,25 +240,23 @@ MM_GCExtensions::computeDefaultMaxHeap(MM_EnvironmentBase *env)
  * internal requirements given that one compilation thread can use up to 256M.
  */
 #define OPENJ9_IN_CGROUP_NATIVE_FOOTPRINT_EXCLUDING_HEAP ((U_64)512 * 1024 * 1024)
-
+		uint64_t cgroupMemory = 0;
 		int32_t rc = omrsysinfo_cgroup_get_memlimit(&cgroupMemory);
 		if (0 == rc) {
 			/* If running in a cgroup with memory limit > 1G, reserve at-least 512M for JVM's internal requirements
 			 * like JIT compilation etc, and extend default max heap memory to at-most 75% of cgroup limit.
 			 */
-			memoryMax = (uintptr_t)OMR_MAX((int64_t)(cgroupMemory / 2), (int64_t)(cgroupMemory - OPENJ9_IN_CGROUP_NATIVE_FOOTPRINT_EXCLUDING_HEAP));
-			memoryMax = (uintptr_t)OMR_MIN(memoryMax, (cgroupMemory / 4) * 3);
+			memoryMax = (uintptr_t)OMR_MAX((int64_t)(usablePhysicalMemory / 2), (int64_t)(usablePhysicalMemory - OPENJ9_IN_CGROUP_NATIVE_FOOTPRINT_EXCLUDING_HEAP));
+			memoryMax = (uintptr_t)OMR_MIN(memoryMax, (usablePhysicalMemory / 4) * 3);
```

Comment:
if we can not remove omrsysinfo_cgroup_get_memlimit() call so I think would be better DO NOT change cgroupMemory to usablePhysicalMemory. In case of using if cgroupMemory it is easy to read - requested and used directly.
Changing it to usablePhysicalMemory means we assume that usablePhysicalMemory must be equal cgroupMemory if rc is 0. (Is this statement is correct? Is there any expanded logic in sysinfo_get_addressable_physical_memory() you want to apply?!). If these two values are equal I would prefer keep code as is (with exception to reduce the scope for cgroupMemory variable)

Reply:
Use of `cgroupMemory` for calculating `memoryMax` is incorrect because it does not take into account address space limit, if any. Value returned by `sysinfo_get_addressable_physical_memory` takes into account cgroup memory (when use of cgroup is enabled) as well as the address space limit.

> Changing it to usablePhysicalMemory means we assume that usablePhysicalMemory must be equal cgroupMemory if rc is 0

This is correct only if there is no address space limit. If there is a limit on address space then `usablePhysicalMemory` would be different than `cgroupMemory` but `rc` would still be 0.

Reply:
Well, it is even worse... if omrsysinfo_cgroup_get_memlimit() does not return proper value looks like problem should be solved on level of Port library. We should resolve this mess first

Reply:
I would say `omrsysinfo_cgroup_get_memlimit` is expected to return the actual memory limit imposed by the cgroup. Correct API to get physical memory is `sysinfo_get_addressable_physical_memory`.
We end up using `omrsysinfo_cgroup_get_memlimit` only to check if there is any cgroup memory limit. I think it would have been more clear if there had been an API like `omrsysinfo_cgroup_is_memlimit_set`. But as of now such an API is not present, so we rely on `omrsysinfo_cgroup_get_memlimit`.

As of now there are no direct users of `omrsysinfo_cgroup_get_memlimit`, but there is use-case for providing cgroup informatoin in javacore in which case I expect this API would be called directly.

Reply:
@charliegracie Do you have suggestion how we can improve this code?

Reply:
Would you please add comment explaining why usablePhysicalMemory should be used instead of cgroupMemory?